File: basic_dsa_one_hundred_seventy_one.java
/**
 * Basic DSA Problems - Binary Tree Maximum Path Sum
 * Max sum of any path (node-to-node, not necessarily through root)
 */
public class basic_dsa_one_hundred_seventy_one {

    // Problem: Return max path sum. Path = any sequence of connected nodes (each node at most once).

    public static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    static int maxSum;

    /**
     * At each node: max path through node = node + max(0, leftGain) + max(0, rightGain).
     * Return to parent: node + max(0, leftGain, rightGain) (single branch for extending path).
     * Time O(n), Space O(h)
     */
    public static int maxPathSum(TreeNode root) {
        maxSum = Integer.MIN_VALUE;
        gain(root);
        return maxSum;
    }

    static int gain(TreeNode node) {
        if (node == null) return 0;
        int L = Math.max(0, gain(node.left));
        int R = Math.max(0, gain(node.right));
        maxSum = Math.max(maxSum, node.val + L + R);
        return node.val + Math.max(L, R);
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Binary Tree Maximum Path Sum");
        System.out.println("=".repeat(60));

        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        System.out.println("Max path sum -> " + maxPathSum(root));
    }
}

File: basic_dsa_one_hundred_seventy_one.cpp
/**
 * Basic DSA - Rotate Array (LeetCode 189)
 * Rotate array to the right by k steps (in-place).
 */
#include <iostream>
#include <vector>
#include <algorithm>

void rotate(std::vector<int>& nums, int k) {
    int n = nums.size();
    k %= n;
    if (k == 0) return;
    std::reverse(nums.begin(), nums.end());
    std::reverse(nums.begin(), nums.begin() + k);
    std::reverse(nums.begin() + k, nums.end());
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Rotate Array\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {1, 2, 3, 4, 5, 6, 7};
    rotate(a1, 3);
    std::cout << "[1,2,3,4,5,6,7], k=3 -> [";
    for (size_t i = 0; i < a1.size(); i++) std::cout << (i ? "," : "") << a1[i];
    std::cout << "]\n";

    std::vector<int> a2 = {-1, -100, 3, 99};
    rotate(a2, 2);
    std::cout << "[-1,-100,3,99], k=2 -> [";
    for (size_t i = 0; i < a2.size(); i++) std::cout << (i ? "," : "") << a2[i];
    std::cout << "]\n";
}
